### PR TITLE
Trace View: Flame graph visualization

### LINKED ISSUE #61383: Trace View: Flame graph

### PR DESCRIPTION
**Pull request by adrapereira**
**What is this feature?**

This PR adds a flame graph as an option to visualise a trace in the Trace View. 

image

When transforming the trace data into a "profile" model that can be read by the flame graph component we:

1. Convert the spans list into a tree using the parent spanId in the `references` field
2. Group spans with the same name at the same level and aggregate their durations
3. Traverse the tree to generate the DataFrame with `levels`, `values`, `self` and `names` arrays

This feature is currently behind the `newTraceView` feature flag. 

**Which issue(s) does this PR fix?**:

Fixes #61383

### CODE REVIEW

File: public/app/plugins/panel/flamegraph/components/TopTable/FlameGraphTopTable.tsx [hunk 1]
```diff
@@ -136,6 +137,14 @@ const FlameGraphTopTable = ({
...
   );
 
+  const handleScroll: React.UIEventHandler = (event) => {
+    const { scrollTop } = event.target as HTMLDivElement;
+
+    if (listRef.current !== null) {
+      listRef.current.scrollTo(scrollTop);
+    }
+  };
+
```
Comment (joey-grafana) [inline public/app/plugins/panel/flamegraph/components/TopTable/FlameGraphTopTable.tsx:147]: Not really sure why this is necessary? Again, the changes in [this PR](https://github.com/grafana/grafana/pull/56555) took a lot into account in terms of scrolling in dashboards/explore - would have to revisit to make sure it's accounted for here :) 

Also must note that these changes will apply to the flamegraph in both the trace view and Phlare/Parca so we ideally have to test at least the trace view and Phlare in both dashboards and explore.

File: public/app/features/explore/TraceView/components/utils/convertTraceToFlameGraph.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      for (let i = 1; i < group.length; i += 1) {
+        res.total += group[i].total;
+      }
+      childrenDur += res.total;
+      return res;
+    });
+    span.total = Math.max(span.duration, childrenDur);
+    span.self = Math.max(0, span.total - childrenDur);
+  }
+  groupSpans(root, 0);
```
Comment (joey-grafana) [inline public/app/features/explore/TraceView/components/utils/convertTraceToFlameGraph.ts:57]: Would be cool to add a test for this and processNode below

**Review by joey-grafana — COMMENTED** [review]
Hey thanks for putting this together 👍 

- Looks like the flame graph/top table is not rendering in Phlare :D 

Screenshot 2023-02-22 at 09 56 37

- When you select an item in the top table the color for the highlighted blocks in the flame graph is black.

Screenshot 2023-02-22 at 09 36 23

- It's interesting that in the trace view the db service is somewhere in the middle, but in the flame graph it's aligned to the left. Is there anything we can do here?

Screenshot 2023-02-22 at 10 10 44
Screenshot 2023-02-22 at 10 10 59

File: public/app/features/explore/TraceView/components/utils/convertTraceToFlameGraph.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      node = spans[span.references[0].spanID] || root;
+    }
+
+    node.children.push(span);
+  });
+
+  // Step 2: group spans with same name
+
+  function groupSpans(span: FlameSpan, d: number) {
+    (span.children || []).forEach((x) => groupSpans(x, d + 1));
```
Comment (aocenas) [inline public/app/features/explore/TraceView/components/utils/convertTraceToFlameGraph.ts:42]: There are recursion limits for all the browsers, even though high https://stackoverflow.com/questions/2805172/what-are-the-js-recursion-limits-for-firefox-chrome-safari-ie-etc I think we could hit them with 100k long traces.

**Comment by aocenas** [comment]
There seems to be lots of cases which I don't see being handled here and not even sure how they should be handled. Even just simple example from our devenv does not produce correct flamegraph:
Screenshot from 2023-02-27 11-11-19
As you can see the second and third level from the bottom should not have the same length. The child is actually longer than the parent which can be problematic to show in flamegraph. Imagine a cases like this:

Screenshot from 2023-02-27 11-29-32
Screenshot from 2023-02-27 11-28-57

Both just does not make that much sense to me in the flamegraph format and not sure how they should be visualized.

**Pull request closed without merging**